compiler/rustc_resolve/src/macros.rs

```diff
@@ -346,9 +347,23 @@ impl<'a> ResolverExpand for Resolver<'a> {
     fn lint_node_id(&mut self, expn_id: ExpnId) -> NodeId {
         // FIXME - make this more precise. This currently returns the NodeId of the
         // nearest closing item - we should try to return the closest parent of the ExpnId
-        self.invocation_parents
-            .get(&expn_id)
-            .map_or(ast::CRATE_NODE_ID, |id| self.def_id_to_node_id[*id])
+        if let Some(mut def_id) = self.invocation_parents.get(&expn_id).copied() {
+            // If a closure resuls from an async desugaring, then it doesn't actually exist in the AST
+            // Lint at its parent instead, which is an actual AST node
```

Comment:
Is it possible to determine that this def-id is "special" without introducing a new table for async closures specifically?
E.g. does it have an entry in `def_id_to_node_id`? Is that entry `DUMMY_NODE_ID`? Or it's just a regular def-id indiscernible from those presenting in AST?

Reply:
Unfortunately, it's just a regular `DefId`, and is actually used during AST lowering and typeck.

An alternative would be to modify the early lint to explicitly check for `Async::Yes` - however, that would make it the only case where we lint a `NodeId` that doesn't really exist in the AST.

Reply:
>An alternative would be to modify the early lint pass to explicitly check for Async::Yes

I think this may actually be a better solution.
The node ids do exist in AST as a part of the `Async::Yes` node, that node is just not yet transformed into a closure.